### ISSUE TITLE
Training

### ISSUE THREAD

**Issue by MeLonJ10** [jspenmar/DeFeat-Net#4]
Hi, it's a exciting job!
In readme, there is only guide about inference. Can you offer me some help to train from scratch?
For example, how to make use of the class "DeFeatLoss".
Thank you so much!

**Comment by jspenmar** [comment]
Hi, thank you for your interest! 
The classes for the network and loss have been implemented such that you can simply pass images from the dataset without worrying about the intermediate steps of combining the different sub-networks and -losses. 

For example, for the [network](https://github.com/jspenmar/DeFeat-Net/blob/master/networks/defeat_net.py#L35)
```
def forward(self, target_frames, support_frames, support_idxs):
    """
    target_frames: (Tensor) (b, 3, h, w) Images we want to predict depth for.
    support_frames: (list[Tensor]) List of support frames (i.e. previous or next frames)
    support_idxs: (list[int]) Index of each support frame wrt the target_frames.
    """
```

In terms of the [loss](https://github.com/jspenmar/DeFeat-Net/blob/master/losses/defeat_loss.py#L38):
```
def forward(self, pred_disp, pred_feats, support_feats, pred_poses, frames, support_frames, gt_depths, k, idxs):
    """
    pred_disp: (Tensor) (b, 3, h, w) Predicted disparity (network output).
    pred_feats: (Tensor) (b, c, h, w) Predicted dense features (network output).
    support_feats: (list[Tensor]) Predicted dense features for the support frames (network output).
    pred_poses: (list[Tensor]) Predicted poses wrt each support frame (network output).
    frames: (Tensor) (b, 3, h, w) Target images we're predicting depth for.
    gt_depths: (Optional[Tensor]) (b, 1, h, w) Ground truth depth (not used at the moment).
    k: (tuple) (b, 4, 4) Camera intrinsics as (k, k_inv).
    idxs: (list[int]) Index of each support frame wrt the target_frames.
    """
```

As you can see, the dataset format is similar to most other monocular depth approaches, so you can use other libraries (e.g. Monodepth) to obtain example dataloaders for Kitti.

Hope this helps!

**Comment by MeLonJ10** [comment]
Hi, 
Would you mind to offer more about the training detail?
For example, learning rate, if the depthnet and posenet pretrained with purely photometric loss or the entire framework is trained from scratch together?
Thanks a lot!

**Comment by jspenmar** [comment]
Hi,

Yes, we use the Adam optimizer, with a base learning rate of 0.001, stepping down by a factor of 10 every 33 epochs (for a total of 100 epochs).
The feature network is trained completely from scratch, while the depth and pose networks use a ResNet encoder pretrained on ImageNet (note that these encoders are optimized separately during training). Both decoders are trained completely from scratch, based only on the photometric/feature losses.

Thank you!
